httpx/middleware/auth.py

```diff
@@ -0,0 +1,257 @@
+import hashlib
+import os
+import re
+import time
+import typing
+from base64 import b64encode
+from collections import defaultdict
+from urllib.request import parse_http_list
+
+from ..exceptions import ProtocolError
+from ..models import AsyncRequest, AsyncResponse, StatusCode
+from ..utils import to_bytes, to_str, unquote
+from .base import BaseMiddleware
+
+
+class BasicAuth(BaseMiddleware):
+    def __init__(
+        self, username: typing.Union[str, bytes], password: typing.Union[str, bytes]
+    ):
+        username = to_bytes(username)
+        password = to_bytes(password)
+        userpass = b":".join((username, password))
+        token = b64encode(userpass).decode().strip()
+
+        self.authorization_header = f"Basic {token}"
+
+    async def __call__(
+        self, request: AsyncRequest, get_response: typing.Callable
+    ) -> AsyncResponse:
+        request.headers["Authorization"] = self.authorization_header
+        return await get_response(request)
+
+
+class CustomAuth(BaseMiddleware):
+    def __init__(self, auth: typing.Callable[[AsyncRequest], AsyncRequest]):
+        self.auth = auth
+
+    async def __call__(
+        self, request: AsyncRequest, get_response: typing.Callable
+    ) -> AsyncResponse:
+        request = self.auth(request)
+        return await get_response(request)
+
+
+class DigestAuth(BaseMiddleware):
+    per_nonce_count: typing.Dict[bytes, int] = defaultdict(lambda: 0)
```

Comment:
A note on this global variable. Requests has a much pragmatic approach to nonce counting, it simply keeps track of the *last* nonce and increases the nonce count if it is the same. However I chose to adhere a bit more closely to the RFC at the potential risk of this structure growing uncontrollably.

I'm not sure what the best solution for this would be. Maybe keep the las N nonce values? Reset on successful authentication?

Reply:
1. Just make it a global. It's actually a benefit for it to be shared across clients, across threads, across tasks.
2. Make it essentially be an LRU dict, storing N items. (1000, 10000, or whatever). Whenever it's more than N items, delete the first entry. (Easy now that python dicts are ordered)
3. We don't need the `_RequestDigestAuth` class. Nonce state stuff can just be a dumb global, and everything can just be implemented on the middleware class.

Reply:
> Just make it a global

It is a mutable class variable so it should qualify as global, do you mean moving it to the module?

> We don't need the _RequestDigestAuth class. Nonce state stuff can just be a dumb global, and everything can just be implemented on the middleware class.

I had this intuition as well, but the `num_401_requests` can't really be global since it would prevent concurrent non-authenticated requests from going through the flow correctly. For instance if I fire 3 concurrent requests the only the first one would go through the flow and the rest would assume the legitimate 401 responses are due to credentials being invalid and return the 401 as the final response.

Following closely the RFC the ideal solution would be to prevent the additional requests to be sent until the first authentication succeeds and then reuse the auth header (which the RFC suggests we should).

However this logic can become even more complex as the server might decide to reject one of the requests and ask for reauthentication so we'd have to block concurrent requests again.

It seemed like quite a complex solution, though there are some benefits, like potentially saving a round trip for authentication but the RFC says the server may or may not take advantage of it. In the end I decided to take the pragmatic approach requests seems to take.

Of course if we decide otherwise I'm happy to implement the full fledged solution.

> Make it essentially be an LRU dict, storing N items. (1000, 10000, or whatever). Whenever it's more than N items, delete the first entry. (Easy now that python dicts are ordered)

Yep, that's what I was thinking as well 👍